crates/primitives-traits/src/transaction/signed.rs

```diff
@@ -179,45 +179,106 @@ impl SignedTransaction for PooledTransaction {
+        }
     }
 
-    fn recover_signer_unchecked_with_buf(
-        &self,
-        buf: &mut Vec<u8>,
-    ) -> Result<Address, RecoveryError> {
-        match self {
-            Self::Legacy(tx) => tx.tx().encode_for_signing(buf),
-            Self::Eip2930(tx) => tx.tx().encode_for_signing(buf),
-            Self::Eip1559(tx) => tx.tx().encode_for_signing(buf),
-            Self::Eip7702(tx) => tx.tx().encode_for_signing(buf),
+    impl SignedTransaction for OpTxEnvelope {
+        fn tx_hash(&self) -> &TxHash {
+            match self {
+                Self::Legacy(tx) => tx.hash(),
+                Self::Eip2930(tx) => tx.hash(),
+                Self::Eip1559(tx) => tx.hash(),
+                Self::Eip7702(tx) => tx.hash(),
+                // TODO: update Deposit variant's matched response to tx.hash_ref() once implemented
+                // in alloys-rs since tx.hash() returns value instead of reference
+                Self::Deposit(tx) => tx.hash(),
+            }
+        }
+
+        fn signature(&self) -> &Signature {
+            match self {
+                Self::Legacy(tx) => tx.signature(),
+                Self::Eip2930(tx) => tx.signature(),
+                Self::Eip1559(tx) => tx.signature(),
+                Self::Eip7702(tx) => tx.signature(),
+                // TODO: The Deposit variant does not have a user signature since it originates
+                // from an L1 event. Decide how to represent or handle signature-less transactions.
+                Self::Deposit(_tx) => unimplemented!(),
+            }
+        }
+
+        fn recover_signer(&self) -> Result<Address, RecoveryError> {
+            let signature_hash = match self {
+                Self::Legacy(tx) => tx.signature_hash(),
+                Self::Eip2930(tx) => tx.signature_hash(),
+                Self::Eip1559(tx) => tx.signature_hash(),
+                Self::Eip7702(tx) => tx.signature_hash(),
+                // TODO: The Deposit variant does not have a user signature since it originates
+                // from an L1 event. Decide how to represent or handle signature-less transactions.
+                Self::Deposit(_tx) => unimplemented!(),
+            };
+
+            recover_signer(self.signature(), signature_hash)
+        }
+
+        fn recover_signer_unchecked_with_buf(
+            &self,
+            buf: &mut Vec<u8>,
+        ) -> Result<Address, RecoveryError> {
+            match self {
+                Self::Legacy(tx) => tx.tx().encode_for_signing(buf),
+                Self::Eip2930(tx) => tx.tx().encode_for_signing(buf),
+                Self::Eip1559(tx) => tx.tx().encode_for_signing(buf),
+                Self::Eip7702(tx) => tx.tx().encode_for_signing(buf),
+                // TODO: The Deposit variant does not have a user signature since it originates
+                // from an L1 event. Decide how to represent or handle signature-less transactions.
+                Self::Deposit(_tx) => unimplemented!(),
```

Comment:
same idea as above to throw error instead

Reply:
same here

https://github.com/paradigmxyz/reth/blob/06453c9553de20fe39e41554f842987bfe9ae650/crates/optimism/primitives/src/transaction/signed.rs#L135-L135

Reply:
updated here as well 🤙 https://github.com/paradigmxyz/reth/blob/94ece0d95fa61e67d9b778d1e39934b3bf6bcfd0/crates/primitives-traits/src/transaction/signed.rs#L265-L280